plutus-core/plc/Main.hs

```diff
@@ -95,7 +93,9 @@ data PrintMode   = Classic | Debug | Readable | ReadableDebug deriving (Show, Re
 type ExampleName = T.Text
 data ExampleMode = ExampleSingle ExampleName | ExampleAvailable
 data EvalMode    = CK | CEK deriving (Show, Read)
-data BudgetMode  = Silent | Verbose ExBudgetMode
+data BudgetMode  = Silent
+                 | forall st. (Eq st, PP.Pretty st, NFData st) =>
+                     Verbose (Cek.ExBudgetMode st PLC.DefaultUni PLC.DefaultFun)
```

Comment:
Now that budgeting modes are extensible we have to specify upfront all the constraints that a budgeting state type has to satisfy, 'cause we won't be able to dispatch on the budgeting state type downstream.

Reply:
Maybe we should export an ADT from the machine with an enumeration of common budgeting functions with the constraints wrapped up, so consumers don't have to deal with this?

Reply:
But it is the consumer who decides what constraints they need. We can't predict everything. Although we can pack constraints, for which we provide instances (`Eq`, `Show`, `Pretty`, `Generic`, `NFData`, smth else?), but I'm not sure if it's helpful.

Reply:
I'm going to leave this for now as is, I think it's not too bad and I don't quite like the idea of packing all known instances together.